go/types/opcache.go

```diff
@@ -27,25 +45,65 @@ func newOpCache(vrw ValueReadWriter) *opCache {
 		WriteBuffer:            1 << 27, // 128MiB
 	})
 	d.Chk.NoError(err, "opening put cache in %s", dir)
-	return &opCache{ops: db, dbDir: dir, vrw: vrw}
+	return &opCacheDb{ops: db, dbDir: dir}
+}
+
+func newOpCache(vrw ValueReadWriter) *opCache {
+	incrementActiveColCount()
+	prefix := [prefixByteSize]byte{}
+	colId := atomic.AddUint32(&collectionId, 1)
+	binary.LittleEndian.PutUint32(prefix[:], colId)
+	return &opCache{vrw: vrw, colId: colId, prefix: prefix[:]}
 }
 
 type opCache struct {
-	ops           *leveldb.DB
-	dbDir         string
-	vrw           ValueReadWriter
-	ldbKeyScratch [1 + hash.ByteLen]byte
-	keyScratch    [initialBufferSize]byte
-	valScratch    [initialBufferSize]byte
+	vrw              ValueReadWriter
+	ldbKeyScratch    [1 + hash.ByteLen]byte
+	keyScratch       [initialBufferSize]byte
+	prefixKeyScratch [initialBufferSize]byte
+	valScratch       [initialBufferSize]byte
+	colId            uint32
+	prefix           []byte
 }
 
 type opCacheIterator struct {
+	opc  *opCache
 	iter iterator.Iterator
 	vr   ValueReader
 }
 
 var uint32Size = binary.Size(uint32(0))
 
+func incrementActiveColCount() {
+	activeMutex.Lock()
+	defer activeMutex.Unlock()
+
+	if activeOpcCnt == 0 {
+		ldb = newLevelDb()
+	}
+	activeOpcCnt++
+}
+
+func decrementActiveColCount() (err error) {
+	activeMutex.Lock()
+	defer activeMutex.Unlock()
+
+	if activeOpcCnt == 1 {
+		d.Chk.NoError(ldb.ops.Close())
+		err = os.RemoveAll(ldb.dbDir)
+		ldb = nil
+	}
+	activeOpcCnt--
+	return
+}
+
+func (p *opCache) PutOp(key, val []byte) error {
+	copy(p.prefixKeyScratch[:], p.prefix)
```

Comment:
You know...I say that Set() can be called from any goroutine, but it definitely isn't safe to call concurrently, because every call to Set() uses the same scratch space buffer. Maybe it's OK to just update the comment for now, but it may be something we need to address later.